### PR TITLE
Fix Instant/Duration math precision & associativity on Windows

### PR DESCRIPTION
**Pull request by bearcage**
**tl;dr** Addition and subtraction on Duration/Instant are not associative on windows because we use the perfcounter frequency in every calculation instead of just when we measure time.

This is my first contrib (PR or Issue) to Rust, so please lmk if I've done this wrong. I followed CONTRIBUTING to the extent I could given my system doesn't seem to be able to build the compiler with changes in the source tree. I also asked about this issue in #rust-beginners a week or so ago, before digging through libstd -- I'm unsure if there's a good way to follow up on that, but I'd be happy to update the docs on the timing structs if this fixes the problem.

## Issue

The `Duration` type keeps seconds in the upper-64 and nanoseconds in the lower-32 bits. In theory doing math on these ought to be basically the same as doing math on any other 64 or 32 bit integral number.

On windows (and I think macos too), however, our math gets messy because the Instant type stores the current point in time in units of HPET Performance Counter counts, not nanoseconds, and does unit conversions on every math operation, rather than just when we measure the time from the system clock.

I tried this code:

```
use std::time::{Duration, Instant};

fn main() {
    let now = Instant::now();
    let offset = Duration::from_millis(5);
    assert_eq!((now + offset) - now, (now - now) + offset);
}
```

On UNIX machines (linux and macos) it behaves as you'd expect -- [no crash](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cf2206c0b7e07d8ecc7767a512364094).

On Windows hosts, however, it blows up because of a precision problem in the Instant +/- Duration math:

```
C:\Users\aberg\work\timetest (master -> origin)
λ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target\debug\timetest.exe`
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `4.999914ms`,
 right: `5ms`', src\main.rs:6:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: process didn't exit successfully: `target\debug\timetest.exe` (exit code: 101)

C:\Users\aberg\work\timetest (master -> origin)
λ cat src\main.rs
use std::time::{Duration, Instant};

fn main() {
    let now = Instant::now();
    let offset = Duration::from_millis(5);
    assert_eq!((now + offset) - now, (now - now) + offset);
}
```

On windows I think this is a consequence of doing the HPET-PerfCounter-Unit conversion on each math operation. I suspect the reason it works on macs is that (from what I could find online) most apple machines report timing in nanoseconds anyway. For anyone interested, the equivalent functions on macos, with a little work to fish out the numerator/denominator from a timebase struct:

* `QueryPerformanceCounter()` -> `mach_absolute_time()`
* `QueryPerformanceFrequency()` -> `mach_timebase_info()`

If this PR ends up working as I expect it to when CI runs the tests, I can make the same changes to the macos implementation.

## Potential Fix

We ought to be able to sort this out by storing nanoseconds, rather than PerfCounter units, that way intermediate math is done in the most precise units we support and we're only doing unit conversions when we actually measure the system clock (and it might even translate to a small perf gain for people doing tons of Instant/Duration math).

I believe this will address the underlying cause of #56034, and make the guessed epsilon constant from #56059 unnecessary. If it's of interest, I can write up how these timing types work on the tier 1 platforms to address #32626 as well, since I'm already in here figuring it out.

## This Patch

To that end, I've got this patch, which I think should fix it on windows, but I'm having trouble testing it -- any time I change anything in libstd I start getting this error, which no amount of clean building seems to resolve:

```
C:\Users\aberg\work\rust (master -> origin)
λ python x.py test --stage 0 --no-doc src/libstd
Updating only changed submodules
Submodules updated in 0.27 seconds
    Finished dev [unoptimized] target(s) in 2.41s
Building stage0 std artifacts (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc)
    Finished release [optimized] target(s) in 6.78s
Copying stage0 std from stage0 (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc / x86_64-pc-windows-msvc)
Building stage0 test artifacts (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc)
   Compiling test v0.0.0 (C:\Users\aberg\work\rust\src\libtest)
error[E0460]: found possibly newer version of crate `std` which `getopts` depends on
  --> src\libtest\lib.rs:36:1
   |
36 | extern crate getopts;
   | ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: perhaps that crate needs to be recompiled?
   = note: the following crate versions were found:
           crate `std`: \\?\C:\Users\aberg\work\rust\build\x86_64-pc-windows-msvc\stage0-sysroot\lib\rustlib\x86_64-pc-windows-msvc\lib\libstd-d7a80ca2ae113c97.rlib
           crate `std`: \\?\C:\Users\aberg\work\rust\build\x86_64-pc-windows-msvc\stage0-sysroot\lib\rustlib\x86_64-pc-windows-msvc\lib\std-d7a80ca2ae113c97.dll
           crate `getopts`: \\?\C:\Users\aberg\work\rust\build\x86_64-pc-windows-msvc\stage0-test\x86_64-pc-windows-msvc\release\deps\libgetopts-ae40a96de5f5d144.rlib

error: aborting due to previous error

For more information about this error, try `rustc --explain E0460`.
error: Could not compile `test`.

To learn more, run the command again with --verbose.
command did not execute successfully: "C:\\Users\\aberg\\work\\rust\\build\\x86_64-pc-windows-msvc\\stage0\\bin\\cargo.exe" "build" "--target" "x86_64-pc-windows-msvc" "-j" "12" "--release" "--manifest-path" "C:\\Users\\aberg\\work\\rust\\src/libtest/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101
failed to run: C:\Users\aberg\work\rust\build\bootstrap\debug\bootstrap test --stage 0 --no-doc src/libstd
Build completed unsuccessfully in 0:00:20
```

---

Since you wrote the linked PRs and might remember looking at related problems:

r? @alexcrichton

### CODE REVIEW

**Review by alexcrichton — COMMENTED** [review]
Thanks so much for the thorough PR and description! I've been feeling for some time now that we probably want to just switch to a bland `Duration` as the internal storage for all `Instant` and possibly `SystemTime`. That I think would also solve this and likely improve our portability of these types across platforms without actually sacrificing much functionality.

One comment I'd have is that we may want to consider the esoteric use case of moving very far back in time. If you take a huge duration and subtract it from the current `Instant` it may work today and may have overflow with this PR. I... think though that this may affect `SystemTime` more than `Instant`, so it may not be a worry for this PR at all! Mind testing this out locally though to see how it goes?

> If it's of interest, I can write up how these timing types work on the tier 1 platforms to address #32626 as well, since I'm already in here figuring it out.

It would definitely be of interest! We'd be more than happy to help take a look at some docs here.

---

cc @sfackler, I think you've had thoughts on the implementations of these APIs on the past, just wanted to make sure you didn't have any objections to this either!

File: src/libstd/sys/windows/time.rs [hunk 1]
```diff
@@ -14,7 +11,7 @@ const INTERVALS_PER_SEC: u64 = NANOS_PER_SEC / 100;
 
 #[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Hash)]
 pub struct Instant {
-    t: c::LARGE_INTEGER,
+    t: u64,
```
[the remark this answers is not in the document]
Comment (bearcage) [reply src/libstd/sys/windows/time.rs]: <3 this idea!

It's basically trading away an extra u32 of memory, per Instant, in exchange for getting better range and same-or-better precision. IMO that's the right call here, since it:

* Eliminates the mental-model difference between Instant and Duration number types — we can treat them like natural numbers, where the distance metric is in the same units as the values and the set bounds are identical between values and distances.
* Lets us normalize the math across all platforms, if we make that change everywhere
* Maybe even lets us simplify this code a fair bit — we might be able to shrink the Instant part of the interface imported from `sys/${platform}/time.rs` down to just a `now` function.

**Comment by bearcage** [comment]
Thanks for the quick turnaround!

I got a fresh shiny Windows VM going after work today, and I can finally get the test suite to run, albeit a bit slowly.

I believe you're correct about the potential for a loss of range with this change as it is today, at least on some machines. Given a pair of windows hosts, if one is giving us PerfCounter increments in nanoseconds and another is giving us microseconds, the precision of our math, and extent of time we can represent, has to vary accordingly.

If you'd like to see one, I can come up with a concrete example of a `Duration` that'll overflow with this patch and wouldn't without it. I'm not sure it'd be a useful test to add to the suite, though, since whether it passes or fails depends on what computer you run it on.

---

Personally I'm leaning toward moving the underlying type to a `Duration`, assuming y'all feel similarly about the extra memory spend being worth the model simplicity we get for it.

I think this probably isn't much of a breaking change since we already don't promise anything about the `Instant` type other than that it represents a point in time (the docs even call it opaque). If this sounds good to y'all, I'll make the change for `sys/windows/time.rs` first, we can land that, and then I'll do the other platforms in libstd, and finally we can see about changing the interface, if it makes sense?

File: src/libstd/sys/windows/time.rs [hunk 2]
```diff
@@ -33,51 +30,45 @@ pub const UNIX_EPOCH: SystemTime = SystemTime {
...
+        Duration::new(diff / NANOS_PER_SEC, (diff % NANOS_PER_SEC) as u32)
     }
 
     pub fn checked_add_duration(&self, other: &Duration) -> Option<Instant> {
-        let freq = frequency() as u64;
-        let t = other.as_secs()
-            .checked_mul(freq)?
-            .checked_add(mul_div_u64(other.subsec_nanos() as u64, freq, NANOS_PER_SEC))?
+        let sum = other.as_secs()
+            .checked_mul(NANOS_PER_SEC)?
```
[the remark this answers is not in the document]
Comment (bearcage) [reply src/libstd/sys/windows/time.rs]: Will do, if we decide to keep the u64 representation. I haven't gotten to play with the u128 type at all yet, so that'll be fun!

**Comment by bearcage** [comment]
In either case, I'll stack on another commit to update the docs with whatever we decide on :-D

File: src/libstd/sys/windows/time.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/libstd/sys/windows/time.rs]: > we might be able to shrink the Instant part of the interface imported from sys/${platform}/time.rs down to just a now function.

Definitely!

**Comment by alexcrichton** [comment]
I think I personally feel that this is the best way to go now. Using a second/nanosecond counter (even if it's not precisely `Duration`) across all platforms I think will make it easier to work with the code internally and externally (less duplication and more consistent internally while also being more portable by having nanosecond resolution externally). 

To be clear I would be worried about something like representing the date 1500 in `SystemTime`. That's pre-epoch on both Unix and Windows, and so representing such a time would require some sort of sign (negative) in `Instant` and/or `SystemTime`. The `Duration` type, however, is unsigned and always represents a positive duration.

I'm not really sure how well this is supported today, however, nor whether it's really a case that comes up that often. I believe for `Instant` it's not really an issue as it's roughly relative to process start. For `SystemTime` it may be a bit worrisome, but anyone storing timestamps so old is likely using a more full-fledged datetime library anyway.

In any case starting with Windows sounds like a good idea, and we can probably just copy all the code into `src/libstd/time.rs` in a future PR! Before going too far down that route though I think it's best to hear from @sfackler first (or at least give them a bit of time to respond)

**Comment by retep998** [comment]
I wonder if there's ever a situation where `QueryPerformanceFrequency` will report a frequency greater than 1,000,000,000...

**Comment by alexcrichton** [comment]
Ah ok, it's been long enough that @bearcage if you want to go ahead and make the changes I think that should be good to do!

**Comment by bearcage** [comment]
Sorry that took me a little to get to — my usual windows machine is still failing to build the test suite (I suspect I've misconfigured visual studio at some point along the way).

Regardless, the last commit on here reworks this PR to use a Duration as the backing type, which simplifies the code a fair bit. If this looks good overall, I'll stack on another commit to reduce the perf_counter module interface to a couple querying functions, and cut another PR for applying the same change to macos.

**Comment by bearcage** [comment]
I'm not sure what's up with that travis failure — looking at [the logs](https://travis-ci.com/rust-lang/rust/jobs/171718864#L1468), it's showing a failure referring to code that doesn't exist in the current HEAD: https://github.com/rust-lang/rust/blob/a31ec05f8d1d865c0e7d447e4f14ce19924b2285/src/libstd/sys/windows/time.rs#L46

Any chance it's got some wonky caching going on?

**Comment by alexcrichton** [comment]
I think that may perhaps be a "merge conflict" of sorts? If you rebase I think that code should show up

**Comment by bearcage** [comment]
Ahh right, I forgot y'all have a fancy post-merge CI system — thanks for the pointer!

Rebased and fixed.

**Comment by alexcrichton** [comment]
Looks great to me!

It's probably worthwhile having a small comment in `struct Instant` explaining what the contained `Duration` is relative to, but other than that I think this is good to go. Do you want to expand the scope though before landing? (aka include other platforms)

**Comment by bearcage** [comment]
Added, thanks!

Let's land this one first, and we can do other platforms separately, then pull 'em all out to a simpler sys interface, if that sounds good to you.

**Comment by alexcrichton** [comment]
> do I need to support these changes through release anywhere or anything like that?

Nope you're all good, after this is merged to master it just sails on to the stable release :)
